### DISCUSSION TITLE
Adding thermostat support for serial controlled MCU

### DISCUSSION THREAD

**Discussion by Derek-K** [arendst/Tasmota#19144] [Ideas]
There is another [project (WThermostatBeca)](https://github.com/klausahrenberg/WThermostatBeca) currently utilize ESP8266/32 as a communication bridge for a "semi" custom thermostat MCU. Interfacing with the MCU is pretty well documented example [here](https://github.com/klausahrenberg/WThermostatBeca/blob/master/docs/BAC-002/MCU-Commands_own_overview.pdf), [here](https://github.com/klausahrenberg/WThermostatBeca/blob/master/docs/MCU-Manual-ENG_Tuya.pdf) and [here](https://github.com/klausahrenberg/WThermostatBeca/blob/master/docs/MCU-Commands%20ET-81.txt), and it is via 9600 baud serial communication.

Tasmota already support for [thermostat](https://tasmota.github.io/docs/Thermostat/), and some of the feature is really nice too (e.g. self learning ramp-up mode) however it is more focus on taking over the complete control of relay and such.

What I am looking for a somewhere in between. I like the Tasmota framework and GUI, which WThermostatBeca lack of. Hoping to use the Tasmota firmware to interface with the MCU via Tx/Rx serial data. I know there is a [serial bridge](https://tasmota.github.io/docs/Serial-to-TCP-Bridge/) feature but it is not what I am looking for.

Perhaps there is a way to use rules that I am not familiar with...
- to add button/slider on the WebGUI
- to send serial commands
- to receive AND decode received serial data

Any suggestions? or that's something completely new that I may have to roll up the sleeve and code it?

**Comment by Jason2866** [comment]
The basic protocol seems is the "standard" Tuya protocol which Tasmota already supports.
The commands which needs to be fired to do the wanted, needs to be done via Rules or Berry.
Start your journey with searching in Tasmota documentation for "Tuya"

Maybe Infos are reusable from this one https://templates.blakadder.com/Moes_WHT-HY609.html

**Reply by Derek-K** [reply]
yes, my bad. I should of finishing up my cup of coffee first... Just found [this](https://tasmota.github.io/docs/Commands/#tuyamcu) as well. And thanks for the link you found. I will take a look and see if I can implement into mine.

What's left is some logics and GUI. May need to upgrade to a ESP32 to use Berry Script...

**Reply by pkkrusty** [reply]
Or use scripting. Probably can work for what you're needing. Berry is a pretty deep pool to jump into if you're starting from zero.

**Reply by Derek-K** [reply]
A bit of experiment I get things kinda hooked up and working. But I run into some issues...

1) Temperature (current & set) displayed on the main page is 5 times cooler. (i.e. when current temp is 25'C it shows 5.0'C. or when set temp is 13'C it shows 2.6C') Is there a 'factor' setting I missed? (I do know the MCU has 0.5'C resolution)
2) In order to use any Tuya command, I have to set GPIO1=Tuya Rx; GPIO3 = Tuya Tx or otherwise Tasmota respond with command unknown (as if Tuya command is not compiled in the binary)

**Reply by Derek-K** [reply]
Taken a crash course on scripting... and here is my proof of concept

```
>D
lock=0
fanspd=1
stmp=20
eco=0
>F
if chg[lock]>0
then
=>TuyaSend1 6,%lock%
endif
if chg[fanspd]>0
then
=>TuyaSend4 103,%(fanspd-1)%
endif
if chg[stmp]>0
then
=>TuyaSend2 2,%(stmp*2)%
endif
if chg[eco]>0
then
=>TuyaSend1 5,%(eco)%
endif
>W
ck(lock "Panel Lock")
bu(eco "Eco Mode ON" "Eco Mode OFF")
pd(fanspd "Fan Speed " "Auto" "High" "Med" "Low")
sl(15 35 stmp "15" "Set Temp" "35")
```

**Reply by Derek-K** [reply]
As for the temperature issue, instead of setting the 'resolution' with decimal place, I suggest to use a multiply factor instead

Original code here
https://github.com/arendst/Tasmota/blob/6382c2806cceac90c6068a613545ded9b9ec3728/tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v2.ino#L597-L614

**Reply by Jason2866** [reply]
Feel free to do a PR, without a breaking change. There is no standard for such stuff. Every manufacturer does what he thinks. We do not maintain such things. It is a fight against windmills.
Tuya is a mess.

**Reply by Derek-K** [reply]
I am not too familiar with the Tasmota code framework layout, trying to figure out the entry point for ``TuyaTempSetRes``, I need to change the parameter to accept float instead of uint8_t before I can make change to ``TuyaAdjustedTemperature()`` (i.e. I can't find where ``Settings->flag2.temperature_resolution`` is being set)

or perhaps without breaking the existing function, to add a new command ``TuyaTempFactor`` which will be something like

```
float TuyaAdjustedTemperature(int16_t packetValue, uint8_t res, float factor) 
 { 
     switch (res) 
     { 
     case 1: 
         return (float)packetValue * factor / 10.0; 
         break; 
     case 2: 
         return (float)packetValue * factor / 100.0; 
         break; 
     case 3: 
         return (float)packetValue * factor / 1000.0; 
         break; 
     default: 
         return (float)packetValue * factor; 
         break; 
     } 
 } 
```

Help? :)

**Reply by barbudor** [reply]
TuyaTempSetRes : https://github.com/arendst/Tasmota/blob/6382c2806cceac90c6068a613545ded9b9ec3728/tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v1.ino#L285
Or if you use V2 (default is still V1) : https://github.com/arendst/Tasmota/blob/6382c2806cceac90c6068a613545ded9b9ec3728/tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v2.ino#L510

Temperature_resolution: https://github.com/arendst/Tasmota/blob/6382c2806cceac90c6068a613545ded9b9ec3728/tasmota/tasmota_support/support_command.ino#L1530

There have been hundreds of discussions around temperature ratio in Tuya because tuya doesn't have a consistent way of handling it. Almost every user comes with a different proposal because his device is handling it differently. So it is likely that any change that may solve your case is going to break for a lot of other people.
Just be very cautious

**Reply by Derek-K** [reply]
@barbudor my next question is... where is ``XdrvMailbox.payload`` coming from and where is ``Settings->flag2.temperature_resolution`` struct defined? And how's the console interpret a command and call ``CmndTemperatureResolution``?

I agree with you, the last thing I want is to have a gang of angry mobsters knocking on my door because I break something :grin: But IMHO, ``TuyaAdjustedTemperature`` should of been more flexible than just setting decimal points to begin with. :smirk:

For now, I figured out my workaround in script
```
...
ctmp=TuyaSNS#Temperature
ctmp=ctmp*5
sctmp=s(2.1ctmp)
stmp=TuyaSNS#TempSet
stmp=stmp*5
sstmp=s(2.1stmp)
>W
Current Temperature: %sctmp%
Set Temperature: %sstmp%
```

And in theory, I could publish my 'adjusted temperature' to MQTT under ``>J``

**Reply by Derek-K** [reply]
Finally I think I figured out changing existing ``TuyaTempSetRes`` command from accepting ``0...3`` to accepting a float value opens up a can of worms.
In short, value ``0...3`` are stored as 2 bits in a structure, and it is also used in ``dtostrfd`` function for formatting. 
Also, it is only applicable for fnID=72 (Temperature Set) and not fnID=71 (Temperature Sensor)

For now, I just added some quick and ugly changes to test my theory and it works with my device. Adding a new command for adjusting factor (default multiply by 1.0) maybe the best option (without breaking existing devices)

**Comment by pkkrusty** [comment]
Not sure you can do `%(stmp*2)%` in scripting. Better to just declare a few new variables and assign them the adjusted value. Your script is short, no memory problem to worry about. Thus:
```
>D
lock=0
fanspd=1
fanspdadjust=0
stmp=20
stmpadjust=0
eco=0
>F
if chg[lock]>0
then
=>TuyaSend1 6,%lock%
endif
if chg[fanspd]>0
then
fanspdadjust=fanspd-1
=>TuyaSend4 103,%fanspdadjust%
endif
if chg[stmp]>0
then
stmpadjust=stmp*2
=>TuyaSend2 2,%stmp%
endif
if chg[eco]>0
then
=>TuyaSend1 5,%eco%
endif
>W
ck(lock "Panel Lock")
bu(eco "Eco Mode ON" "Eco Mode OFF")
pd(fanspd "Fan Speed " "Auto" "High" "Med" "Low")
sl(15 35 stmp "15" "Set Temp" "35")
```

**Reply by Derek-K** [reply]
Thanks. I have tested my script, ``%(fanspd-1)%`` and ``%(stmp*2)%`` are totally fine. Next, I need to figure out how to decode Tuya responds (at least during init to populate the variables)

Worst case I guess I can try to grab the data via MQTT over the network, but it would be better if I can retrieve the data internally.
